flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java

```diff
@@ -0,0 +1,184 @@
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    private boolean matchRow(RowData row, RowData mayMatch) {
+        for (int i = 0; i < rowType.getFieldCount(); ++i) {
+            if (!Objects.equals(
+                    fieldGetters[i].getFieldOrNull(row),
+                    fieldGetters[i].getFieldOrNull(mayMatch))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void prepare() {
+        if (!prepared) {
+            synchronized (this) {
+                if (!prepared) {
+                    doPrepare();
+                    prepared = true;
+                }
+            }
+        }
+    }
+
+    private void doPrepare() {
+        this.dataMap = new HashMap<>();
+        if (isFiltering) {
+            this.fieldGetters =
+                    IntStream.range(0, rowType.getFieldCount())
+                            .mapToObj(i -> RowData.createFieldGetter(rowType.getTypeAt(i), i))
+                            .toArray(RowData.FieldGetter[]::new);
+
+            TypeSerializer<RowData> serializer = typeInfo.createSerializer(new ExecutionConfig());
+            for (byte[] bytes : serializedData) {
+                try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+                        DataInputViewStreamWrapper inView = new DataInputViewStreamWrapper(bais)) {
+                    RowData partition = serializer.deserialize(inView);
+                    List<RowData> partitions =
+                            dataMap.computeIfAbsent(hash(partition), k -> new ArrayList<>());
+                    partitions.add(partition);
+                } catch (Exception e) {
+                    throw new TableException("Unable to deserialize the value.", e);
+                }
+            }
+        }
+    }
+
+    private int hash(RowData row) {
+        return Objects.hash(Arrays.stream(fieldGetters).map(g -> g.getFieldOrNull(row)).toArray());
+    }
+
+    public static boolean isEqual(DynamicFilteringData data, DynamicFilteringData another) {
+        if (data == null) {
+            return another == null;
+        }
+        if (another == null
+                || (data.isFiltering != another.isFiltering)
+                || !data.typeInfo.equals(another.typeInfo)
+                || !data.rowType.equals(another.rowType)
+                || data.serializedData.size() != another.serializedData.size()) {
+            return false;
+        }
+
+        BytePrimitiveArrayComparator comparator = new BytePrimitiveArrayComparator(true);
+        for (int i = 0; i < data.serializedData.size(); i++) {
+            if (comparator.compare(data.serializedData.get(i), another.serializedData.get(i))
+                    != 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @VisibleForTesting
+    public Collection<RowData> getData() {
```

Comment:
this method is not only used for testing, but also can be used for advanced filtering which `contains` methods can not meet the requirement

Reply:
So far the method is not used in formal code, including the dynamic filtering implementation of hive source. I'd prefer to remain the annotation for now. We can remove it when it's actually used.

Reply:
I think it should be a public interface, some lake house connectors has statistics info in partition spec, the partition can be skipped not only based on the partition value but also base on the statistics. If these connectors want to support dpp, they will not change the flink framework code.